net/FlatBuffers/ByteBuffer.cs

```diff
@@ -50,12 +54,16 @@ namespace FlatBuffers
 {
     public abstract class ByteBufferAllocator : IDisposable
     {
-#if UNSAFE_BYTEBUFFER
-        public unsafe byte* Buffer
-        {
-            get;
-            protected set;
-        }
+#if ENABLE_SPAN_T
+        public abstract Span<byte> GetSpan(int start, int length);
+        public abstract ReadOnlySpan<byte> GetReadOnlySpan(int start, int length);
+        public abstract Memory<byte> GetMemory(int start, int length);
+        public abstract ReadOnlyMemory<byte> GetReadOnlyMemory(int start, int length);
+
+        public virtual Span<byte> GetSpan() { return GetSpan(0, Length); }
```

Comment:
Maybe we should make this non virtual to avoid one additional virtual call?  Not sure when the base implementation would not be enough.

Reply:
I bounced back and forth on this... But I ended up here "in the name of performance".

The default `ByteArrayAllocator` class overrides these methods because I saw an extremely small performance gain in doing so. I never verified exactly why, but my assumption is the performance gain is caused by not having to do bounds checks on the array. If you just call `GetSpan()` or `GetSpan(start)`, we directly return `array` and `array.AsSpan(start)`, thus no bounds checks for the first call, and only one bounds check for the second call.

I can go back to making these non-virtual, as I think it cleans up the API. But that was the reasoning I had these `virtual`.

Reply:
If you can get a perf gain by overriding then I would be inclined to leave it as is.

Reply:
>  If you just call GetSpan() or GetSpan(start), we directly return array and array.AsSpan(start), thus no bounds checks for the first call, and only one bounds check for the second call

I'd have thought you could get the equivalent of that by having:
```C#
public abstract Span<byte> GetSpan();
public Span<byte> GetSpan(int start) => GetSpan().Slice(start);
public Span<byte> GetSpan(int start, int length) => GetSpan().Slice(start, length);
```
Does that not perform as well?

Reply:
The three benchmarks I'm adding to the repo showed a small improvement doing it the way I currently have it. Let me get some numbers again to make sure.

Reply:
(And if you did do that, then the API could even just expose `GetSpan()`, parameterless, and let the caller just do their own slicing.)

Reply:
(And it could then just be a property, since you wouldn't need to worry about overloads :smile:)

Reply:
I ran the benchmarks 3 times with each way, and I didn't see any noticeable perf differences.

## Current code:

|             Method |      Mean |     Error |    StdDev |    Median | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------- |----------:|----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| BuildNestedMonster | 11.964 ms | 0.1665 ms | 0.1476 ms | 11.963 ms |   4609.3750 |           - |           - |            18.46 MB |
|       BuildMonster |  5.849 ms | 0.1155 ms | 0.2678 ms |  5.724 ms |   2054.6875 |           - |           - |             8.24 MB |
|         TestTables | 86.696 ms | 1.6580 ms | 1.4697 ms | 86.228 ms |    833.3333 |    833.3333 |    833.3333 |               32 MB |

|             Method |      Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------- |----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| BuildNestedMonster | 11.886 ms | 0.2122 ms | 0.1985 ms |   4609.3750 |           - |           - |            18.46 MB |
|       BuildMonster |  5.652 ms | 0.0884 ms | 0.0784 ms |   2054.6875 |           - |           - |             8.24 MB |
|         TestTables | 90.610 ms | 2.6365 ms | 3.1385 ms |    833.3333 |    833.3333 |    833.3333 |               32 MB |

|             Method |      Mean |     Error |    StdDev |    Median | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------- |----------:|----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| BuildNestedMonster | 11.805 ms | 0.1513 ms | 0.1415 ms | 11.843 ms |   4609.3750 |           - |           - |            18.46 MB |
|       BuildMonster |  5.880 ms | 0.1545 ms | 0.3122 ms |  5.751 ms |   2054.6875 |           - |           - |             8.24 MB |
|         TestTables | 88.449 ms | 0.6658 ms | 0.6228 ms | 88.661 ms |    833.3333 |    833.3333 |    833.3333 |               32 MB |

## Suggested

|             Method |      Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------- |----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| BuildNestedMonster | 11.836 ms | 0.1115 ms | 0.0988 ms |   4609.3750 |           - |           - |            18.46 MB |
|       BuildMonster |  5.585 ms | 0.1064 ms | 0.0995 ms |   2054.6875 |           - |           - |             8.24 MB |
|         TestTables | 91.392 ms | 0.8878 ms | 0.7870 ms |    833.3333 |    833.3333 |    833.3333 |               32 MB |

|             Method |      Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------- |----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| BuildNestedMonster | 12.159 ms | 0.1445 ms | 0.1281 ms |   4609.3750 |           - |           - |            18.46 MB |
|       BuildMonster |  5.668 ms | 0.0386 ms | 0.0361 ms |   2054.6875 |           - |           - |             8.24 MB |
|         TestTables | 89.481 ms | 0.6771 ms | 0.6333 ms |    833.3333 |    833.3333 |    833.3333 |               32 MB |

|             Method |      Mean |     Error |    StdDev |    Median | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------- |----------:|----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| BuildNestedMonster | 12.448 ms | 0.2476 ms | 0.6695 ms | 12.131 ms |   4609.3750 |           - |           - |            18.46 MB |
|       BuildMonster |  5.578 ms | 0.1005 ms | 0.0940 ms |  5.550 ms |   2054.6875 |           - |           - |             8.24 MB |
|         TestTables | 89.953 ms | 1.4435 ms | 1.2796 ms | 89.771 ms |    833.3333 |    833.3333 |    833.3333 |               32 MB |

So I refactored to the suggested 4 abstract properties. (Honestly, I kind of forgot you can have a `Span` property on a class ... just not a field.)

Then I went and implemented some of the other suggestions in this PR, and the new numbers I have look even better (I assume it is mostly due to the bounds-check eliding in the `Put` method):

|             Method |      Mean |     Error |    StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------- |----------:|----------:|----------:|------------:|------------:|------------:|--------------------:|
| BuildNestedMonster | 11.531 ms | 0.1855 ms | 0.1644 ms |   4609.3750 |           - |           - |            18.46 MB |
|       BuildMonster |  5.567 ms | 0.0707 ms | 0.0661 ms |   2054.6875 |           - |           - |             8.24 MB |
|         TestTables | 85.469 ms | 1.0523 ms | 0.9843 ms |    833.3333 |    833.3333 |    833.3333 |               32 MB |